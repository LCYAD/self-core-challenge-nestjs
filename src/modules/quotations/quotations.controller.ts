import {
  Body,
  Controller,
  Get,
  Param,
  ParseArrayPipe,
  Post
} from '@nestjs/common'
import { validationExceptionFactory } from '../../utils/validation.util'
import { QuotationIdParamDto } from '../../dtos/quotation/id.param.dto'
import { QuotationsPostReqDto } from './dtos/quotationsPostReq.dto'
import { QuotationHandlerService } from './quotations.handler.service'

@Controller('quotations')
export class QuotationsController {
  constructor(
    private readonly quotationHandlerService: QuotationHandlerService
  ) {}

  @Get(':id')
  getQuotation(@Param() param: QuotationIdParamDto) {
    return this.quotationHandlerService.getQuotationById(param)
  }

  @Post()
  postQuotations(
    @Body(
      new ParseArrayPipe({
        items: QuotationsPostReqDto,
        exceptionFactory: validationExceptionFactory
      })
    )
    body: QuotationsPostReqDto[]
  ) {
    return this.quotationHandlerService.createQuotations(body)
  }
}
