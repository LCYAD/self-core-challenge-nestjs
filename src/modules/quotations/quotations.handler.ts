import { Injectable } from '@nestjs/common'

import _ from 'lodash'
import { customAlphabet } from 'nanoid'

import type { QuotationBaseDto } from '@dtos/quotation/base.dto'
import { QuotationIdParamDto } from '@dtos/quotation/id.param.dto'
import { ResourcesNotFoundException } from '@exceptions/resourcesNotFound.exception'
import { redis } from '@utils/redis.util'

import type { QuotationsPostReqDto } from './dtos/quotationsPostReq.dto'

@Injectable()
export class QuotationsHandler {
  async getQuotationById({ id }: QuotationIdParamDto) {
    const quotation = await redis.get(id)
    if (!quotation) {
      throw new ResourcesNotFoundException(
        'QuotationsHandler:getQuotationById',
        'Quotation not found'
      )
    }
    return JSON.parse(quotation)
  }

  async createQuotations(body: QuotationsPostReqDto[]) {
    const now = Date.now()
    const duration = 5 * 60 * 1000
    const expireAt = new Date(now + duration).toISOString()
    const generateQuotationId = () => customAlphabet('0123456789', 12)()
    const quotations = _.map(body, (quotation) => ({
      ...quotation,
      quotationId: generateQuotationId(),
      expireAt
    }))
    await Promise.all(
      _.map(quotations, (quotation: QuotationBaseDto) =>
        redis.set(quotation.quotationId, JSON.stringify(quotation))
      )
    )
    return quotations
  }
}
