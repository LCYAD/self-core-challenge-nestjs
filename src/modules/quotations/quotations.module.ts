import { Module } from '@nestjs/common'
import { QuotationsController } from './quotations.controller'
import { QuotationHandlerService } from './quotations.handler.service'

@Module({
  imports: [],
  controllers: [QuotationsController],
  providers: [QuotationHandlerService]
})
export class QuotationsModule {}
