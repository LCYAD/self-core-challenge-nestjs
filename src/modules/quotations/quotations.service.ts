import { Injectable } from '@nestjs/common'
import { QUOTATION } from '../../constants/fakers/quotation'

@Injectable()
export class QuotationHandlerService {
  getQuotationById() {
    return QUOTATION
  }
}
