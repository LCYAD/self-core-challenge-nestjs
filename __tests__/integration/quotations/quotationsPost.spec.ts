import { HttpStatus } from '@nestjs/common'
import {
  FastifyAdapter,
  NestFastifyApplication
} from '@nestjs/platform-fastify'

import _ from 'lodash'
import nano from 'nanoid'

import { COMMON_FIELD } from '@constants/fakers/common.faker.constant'
import { quotationsPostReqValidExample } from '@modules/quotations/dtos/quotationsPostReq.dto'
import { QuotationsHandler } from '@modules/quotations/quotations.handler'
import { QuotationsModule } from '@modules/quotations/quotations.module'
import {
  generateSequentialTests,
  mockGenericApp
} from '@tests/helper/e2e.helper'
import { redis } from '@utils/redis.util'

jest.mock('ioredis', () => require('ioredis-mock'))

const path = '/quotations'

describe(`POST ${path}`, () => {
  let app: NestFastifyApplication
  let quotationsHandler: QuotationsHandler
  let mockFunctionGroup

  beforeAll(async () => {
    jest.spyOn(Date, 'now').mockReturnValue(COMMON_FIELD.CURRENT_TIME_MILLI_SEC)
    jest
      .spyOn(nano, 'customAlphabet')
      .mockImplementation(() => () => COMMON_FIELD.QUOTATION_ID_FIELD as string)
    const appModule: NestFastifyApplication = (
      await mockGenericApp({
        module: {
          imports: [QuotationsModule]
        }
      })
    ).createNestApplication(new FastifyAdapter())

    quotationsHandler = appModule.get(QuotationsHandler)
    mockFunctionGroup = {
      spyHandlerCreateQuotations: jest
        .spyOn(quotationsHandler, 'createQuotations')
        .mockName('spyHandlerCreateQuotations'),
      spyRedisSet: jest.spyOn(redis, 'set').mockImplementation()
    }

    app = await appModule.init()
  })

  describe('success cases', () => {
    it('should return correct response', async () => {
      const response = await app.inject({
        method: 'POST',
        url: path,
        payload: [quotationsPostReqValidExample]
      })
      expect(response.statusCode).toEqual(HttpStatus.CREATED)
      generateSequentialTests({
        mockFunctionGroup,
        snapshots: ['spyHandlerCreateQuotations', 'spyRedisSet'],
        apiResponse: response
      })
    })
    it('should return correct response if coordinates is not provided', async () => {
      const response = await app.inject({
        method: 'POST',
        url: path,
        payload: [
          {
            ...quotationsPostReqValidExample,
            stops: [
              _.omit(quotationsPostReqValidExample.stops[0], ['coordinates']),
              _.omit(quotationsPostReqValidExample.stops[1], ['coordinates'])
            ]
          }
        ]
      })
      expect(response.statusCode).toEqual(HttpStatus.CREATED)
      generateSequentialTests({
        mockFunctionGroup,
        snapshots: ['spyHandlerCreateQuotations', 'spyRedisSet'],
        apiResponse: response
      })
    })
  })

  describe('error cases', () => {
    it('should throw Validation Error if body is invalid', async () => {
      const response = await app.inject({
        method: 'POST',
        url: path,
        payload: [_.omit(quotationsPostReqValidExample, ['deliveryBy'])]
      })
      expect(response.statusCode).toEqual(HttpStatus.UNPROCESSABLE_ENTITY)
      generateSequentialTests({
        mockFunctionGroup,
        notCalled: ['spyHandlerCreateQuotations', 'spyRedisSet'],
        apiResponse: response
      })
    })
    it('should throw Validation Error if scheduleAt is after deliveryBy ', async () => {
      const response = await app.inject({
        method: 'POST',
        url: path,
        payload: [
          {
            ...quotationsPostReqValidExample,
            scheduleAt: COMMON_FIELD.AFTER_1HR_CURRENT_TIME,
            deliveryBy: COMMON_FIELD.CURRENT_TIME
          }
        ]
      })
      expect(response.statusCode).toEqual(HttpStatus.UNPROCESSABLE_ENTITY)
      generateSequentialTests({
        mockFunctionGroup,
        notCalled: ['spyHandlerCreateQuotations', 'spyRedisSet'],
        apiResponse: response
      })
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  afterAll(async () => {
    await app.close()
  })
})
