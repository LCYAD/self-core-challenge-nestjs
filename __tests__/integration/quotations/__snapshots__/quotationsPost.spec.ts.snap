// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /quotations error cases should throw Validation Error if body is invalid 1`] = `
{
  "code": "10001",
  "detail": [
    {
      "msg": "{"IsLaterThan":"deliveryBy must be later than scheduleAt, now","isDateString":"deliveryBy must be a valid ISO 8601 date string"}",
      "path": "0.deliveryBy",
    },
  ],
  "type": "VALIDATION_ERROR",
}
`;

exports[`POST /quotations error cases should throw Validation Error if scheduleAt is after deliveryBy  1`] = `
{
  "code": "10001",
  "detail": [
    {
      "msg": "{"IsLaterThan":"deliveryBy must be later than scheduleAt, now"}",
      "path": "0.deliveryBy",
      "value": "2022-05-21T00:00:00.000Z",
    },
  ],
  "type": "VALIDATION_ERROR",
}
`;

exports[`POST /quotations success cases should return correct response 1`] = `
[MockFunction spyHandlerCreateQuotations] {
  "calls": [
    [
      [
        {
          "deliveryBy": "2022-05-21T01:00:00.000Z",
          "item": {
            "categories": [
              "test",
            ],
            "handlingInstructions": [
              "test",
            ],
            "quantity": 1,
            "weight": 1,
          },
          "location": "Hong Kong",
          "scheduleAt": "2022-05-21T00:05:00.000Z",
          "stops": [
            {
              "address": "999 Prince Edward Road",
              "coordinates": {
                "lat": "22.335313",
                "lng": "114.175840",
              },
            },
            {
              "address": "InnoCentre, 72 Tat Chee Avenue",
              "coordinates": {
                "lat": "22.335313",
                "lng": "114.175840",
              },
            },
          ],
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`POST /quotations success cases should return correct response 2`] = `
[MockFunction] {
  "calls": [
    [
      "182379172831",
      "{"scheduleAt":"2022-05-21T00:05:00.000Z","deliveryBy":"2022-05-21T01:00:00.000Z","location":"Hong Kong","stops":[{"coordinates":{"lat":"22.335313","lng":"114.175840"},"address":"999 Prince Edward Road"},{"coordinates":{"lat":"22.335313","lng":"114.175840"},"address":"InnoCentre, 72 Tat Chee Avenue"}],"item":{"quantity":1,"weight":1,"categories":["test"],"handlingInstructions":["test"]},"quotationId":"182379172831","expireAt":"2022-05-21T00:05:00.000Z"}",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`POST /quotations success cases should return correct response 3`] = `
[
  {
    "deliveryBy": "2022-05-21T01:00:00.000Z",
    "expireAt": "2022-05-21T00:05:00.000Z",
    "item": {
      "categories": [
        "test",
      ],
      "handlingInstructions": [
        "test",
      ],
      "quantity": 1,
      "weight": 1,
    },
    "location": "Hong Kong",
    "quotationId": "182379172831",
    "scheduleAt": "2022-05-21T00:05:00.000Z",
    "stops": [
      {
        "address": "999 Prince Edward Road",
        "coordinates": {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
      {
        "address": "InnoCentre, 72 Tat Chee Avenue",
        "coordinates": {
          "lat": "22.335313",
          "lng": "114.175840",
        },
      },
    ],
  },
]
`;

exports[`POST /quotations success cases should return correct response if coordinates is not provided 1`] = `
[MockFunction spyHandlerCreateQuotations] {
  "calls": [
    [
      [
        {
          "deliveryBy": "2022-05-21T01:00:00.000Z",
          "item": {
            "categories": [
              "test",
            ],
            "handlingInstructions": [
              "test",
            ],
            "quantity": 1,
            "weight": 1,
          },
          "location": "Hong Kong",
          "scheduleAt": "2022-05-21T00:05:00.000Z",
          "stops": [
            {
              "address": "999 Prince Edward Road",
              "coordinates": undefined,
            },
            {
              "address": "InnoCentre, 72 Tat Chee Avenue",
              "coordinates": undefined,
            },
          ],
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`POST /quotations success cases should return correct response if coordinates is not provided 2`] = `
[MockFunction] {
  "calls": [
    [
      "182379172831",
      "{"scheduleAt":"2022-05-21T00:05:00.000Z","deliveryBy":"2022-05-21T01:00:00.000Z","location":"Hong Kong","stops":[{"address":"999 Prince Edward Road"},{"address":"InnoCentre, 72 Tat Chee Avenue"}],"item":{"quantity":1,"weight":1,"categories":["test"],"handlingInstructions":["test"]},"quotationId":"182379172831","expireAt":"2022-05-21T00:05:00.000Z"}",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`POST /quotations success cases should return correct response if coordinates is not provided 3`] = `
[
  {
    "deliveryBy": "2022-05-21T01:00:00.000Z",
    "expireAt": "2022-05-21T00:05:00.000Z",
    "item": {
      "categories": [
        "test",
      ],
      "handlingInstructions": [
        "test",
      ],
      "quantity": 1,
      "weight": 1,
    },
    "location": "Hong Kong",
    "quotationId": "182379172831",
    "scheduleAt": "2022-05-21T00:05:00.000Z",
    "stops": [
      {
        "address": "999 Prince Edward Road",
      },
      {
        "address": "InnoCentre, 72 Tat Chee Avenue",
      },
    ],
  },
]
`;
